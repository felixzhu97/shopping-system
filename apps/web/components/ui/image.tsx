"use client"

import * as React from "react"
import { cn } from "@/lib/utils"
import NextImage from "next/image"

export interface ImageProps extends React.ComponentPropsWithoutRef<"img"> {
  fallbackSrc?: string
  fallbackAlt?: string
  wrapperClassName?: string
}

export const Image = React.forwardRef<HTMLImageElement, ImageProps>(
  ({ className, src, alt = "", fallbackSrc, fallbackAlt, wrapperClassName, ...props }, ref) => {
    const [error, setError] = React.useState(false)
    
    const handleError = () => {
      setError(true)
    }
    
    return (
      <div className={cn("relative overflow-hidden", wrapperClassName)}>
        {error || !src ? (
          <img
            className={cn("object-cover transition-opacity", className)}
            src={fallbackSrc || `/placeholder.svg?height=300&width=300&text=${fallbackAlt || alt}`}
            alt={fallbackAlt || alt}
            ref={ref}
            {...props}
          />
        ) : (
          <img
            className={cn("object-cover transition-opacity", className)}
            src={src}
            alt={alt}
            onError={handleError}
            ref={ref}
            {...props}
          />
        )}
      </div>
    )
  }
)

Image.displayName = "Image"

export default Image 