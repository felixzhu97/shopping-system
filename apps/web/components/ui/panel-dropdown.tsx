import React, { useEffect, useRef } from 'react';
import { cn } from '@/lib/utils';

interface PanelDropdownProps {
  open: boolean;
  onClose: () => void;
  className?: string;
  style?: React.CSSProperties;
  children: React.ReactNode;
  containerClassName?: string;
  heightClassName?: string;
}

/**
 * 通用下拉面板组件，支持吸顶、动画、点击外部关闭。
 * 用于搜索栏、购物车等下拉弹窗。
 */
export default function PanelDropdown({
  open,
  onClose,
  className = '',
  style = {},
  children,
  containerClassName = '',
  heightClassName = '',
}: PanelDropdownProps) {
  const ref = useRef<HTMLDivElement>(null);

  // 点击外部关闭
  useEffect(() => {
    if (!open) return;
    function handleClick(e: MouseEvent) {
      if (ref.current && !ref.current.contains(e.target as Node)) {
        onClose();
      }
    }
    document.addEventListener('mousedown', handleClick);
    return () => document.removeEventListener('mousedown', handleClick);
  }, [open, onClose]);

  return (
    <div
      className={cn(
        'fixed inset-x-0 top-12 z-40 overflow-hidden transition-all duration-300 ease-in-out',
        open ? `opacity-100 ${heightClassName}` : 'h-0 opacity-0',
        containerClassName
      )}
      style={style}
    >
      <div
        ref={ref}
        className={cn(
          'bg-[rgba(251,251,253,0.95)] backdrop-blur-md border-b border-gray-200 shadow-sm h-full',
          className
        )}
      >
        {children}
      </div>
    </div>
  );
}
