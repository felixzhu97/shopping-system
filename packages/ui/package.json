{
  "name": "ui",
  "version": "0.0.0",
  "private": true,
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "license": "MIT",
  "scripts": {
    "lint": "eslint .",
    "build": "tsup ./src/index.ts --format esm,cjs --dts --external react",
    "dev": "tsup ./src/index.ts --format esm,cjs --watch --dts --external react",
    "clean": "rm -rf dist && rm -rf node_modules"
  },
  "dependencies": {
    "react": "^19"
  },
  "devDependencies": {
    "@types/react": "^19",
    "eslint": "^8",
    "eslint-config-next": "15.2.4",
    "tsup": "^8.0.2",
    "typescript": "^5"
  }
}
