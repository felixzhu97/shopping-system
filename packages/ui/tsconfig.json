{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "declaration": true,
    "jsx": "react-jsx",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["src"],
  "exclude": ["node_modules", "dist"]
}
