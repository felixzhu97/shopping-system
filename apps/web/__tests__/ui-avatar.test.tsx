import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Avatar, AvatarFallback, AvatarImage } from '../components/ui/avatar';

describe('Avatar Components', () => {
  it('should render Avatar correctly', () => {
    const { container } = render(
      <Avatar>
        <AvatarImage src="/test-avatar.jpg" alt="Test" />
        <AvatarFallback>AB</AvatarFallback>
      </Avatar>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render AvatarImage', () => {
    const { container } = render(
      <Avatar>
        <AvatarImage src="/test-avatar.jpg" alt="Test Avatar" />
      </Avatar>
    );
    // Check if Avatar container is rendered
    expect(container.firstChild).toBeInTheDocument();
  });

  it('should render AvatarFallback', () => {
    render(
      <Avatar>
        <AvatarFallback>JD</AvatarFallback>
      </Avatar>
    );
    expect(screen.getByText('JD')).toBeInTheDocument();
  });

  it('should show fallback when image fails to load', () => {
    render(
      <Avatar>
        <AvatarImage src="/non-existent.jpg" alt="Test" />
        <AvatarFallback>FB</AvatarFallback>
      </Avatar>
    );

    // Both should exist in DOM, but fallback will be shown when image fails
    expect(screen.getByText('FB')).toBeInTheDocument();
  });

  it('should apply custom className', () => {
    const { container } = render(
      <Avatar className="custom-class">
        <AvatarFallback>CC</AvatarFallback>
      </Avatar>
    );
    expect(container.firstChild).toHaveClass('custom-class');
  });
});
