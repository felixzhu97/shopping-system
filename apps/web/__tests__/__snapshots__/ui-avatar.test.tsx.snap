// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar Components > should render Avatar correctly 1`] = `
<div>
  <span
    class="relative flex h-10 w-10 shrink-0 overflow-hidden rounded-full"
  >
    <span
      class="flex h-full w-full items-center justify-center rounded-full bg-muted"
    >
      AB
    </span>
  </span>
</div>
`;
